src/librustc/ty/sty.rs

```diff
@@ -1472,6 +1472,43 @@ impl<'a, 'gcx, 'tcx> TyS<'tcx> {
         }
     }
 
+    pub fn conservative_is_uninhabited(&self, tcx: TyCtxt<'a, 'gcx, 'tcx>) -> bool {
+        // Checks whether a type is definitely uninhabited. This is
+        // conservative: for some types that are uninhabited we return `false`,
+        // but we only return `true` for types that are definitely uninhabited.
+        match self.sty {
+            ty::Never => true,
+            ty::Adt(def, _) => {
+                // Any ADT is uninhabited if:
+                // (a) It has no variants (i.e. an empty `enum`);
+                // (b) Each of its variants (a single one in the case of a `struct`) has at least
+                //     one uninhabited field.
```

Comment:
What if this is an `union`?

Reply:
The same logic should apply as far as I can tell?

Reply:
Good catch @eddyb, I forgot that unions were considered ADTs. I don't think the same logic applies. Or at minimum this is up for debate. 

In the terms from @RalfJung's [Two Kinds of Invariants](https://www.ralfj.de/blog/2018/08/22/two-kinds-of-invariants.html) blog post, the justification for optimizations around uninhabitedness arises from the **validity invariant** (to a large extent). All values that the compiler thinks are initialized must meet this invariant: presumably this includes the return value. This means that, yes, if we find a `!` embedded within a struct or tuple, we can justify saying that the fn never returns (etc).

But the validity invariant for `&!` is different, and doesn't require that the pointer is valid. The same is true of unions. 

We might wind up requiring a stronger invariant of fn return values, mind you, that would make returning `&!` UB, but I think this is unlikely and certainly not decided.

Reply:
Yes please exclude unions. (Sorry should have read this before adding my other comment.)